pubsub_test.go

```diff
@@ -2,12 +2,53 @@ package pubsub
 
 import (
 	"context"
+	"fmt"
+	"os"
 	"testing"
 	"time"
 )
 
+func TestMain(m *testing.M) {
+	// For all tests to work reliably in parallel,
+	// we require being able to open many files at a time.
+	// Many environments have lower limits,
+	// such as Linux's "ulimit -n" soft limit defaulting to 1024.
+	// On a machine with 16 CPU threads, 8k seems to be enough.
+	const wantFileLimit = 8 << 10
+	files := make([]*os.File, 0, wantFileLimit)
+	for i := 0; i < wantFileLimit; i++ {
+		file, err := os.Open("pubsub_test.go")
+		if err != nil {
+			if i == 0 {
+				panic(err) // the file doesn't exist?
+			}
+			fmt.Fprintf(os.Stderr, "Skipping parallel runs of tests; open file limit %d is below %d.\n",
+				i, wantFileLimit)
+			fmt.Fprintf(os.Stderr, "On Linux, consider running: ulimit -Sn %d\n",
+				wantFileLimit*2)
+			canRunInParallel = false
+			break
+		}
+		files = append(files, file)
+	}
+	for _, file := range files {
+		file.Close()
+	}
+	os.Exit(m.Run())
+}
+
+var canRunInParallel = true
```

Comment:
should default to false.

Reply:
The checking code always runs, does it make a difference? :)